rosidl_generator_c/resource/msg__functions.c.em

```diff
@@ -471,17 +471,16 @@ bool
     if (!data) {
       return false;
     }
+    output->data = data;
     for (size_t i = output->capacity; i < input->size; ++i) {
-      if (!@(message_typename)__init(&data[i])) {
+      if (!@(message_typename)__init(&output->data[i])) {
         /* free currently allocated and return false */
         for (; i-- > output->capacity; ) {
-          @(message_typename)__fini(&data[i]);
+          @(message_typename)__fini(&output->data[i]);
         }
-        allocator.deallocate(data, allocator.state);
         return false;
       }
     }
-    output->data = data;
     output->capacity = input->size;
   }
   output->size = input->size;
```

Comment:
This isn't directly related to what you are fixing here, but should we move this *after* the loop below that copies the individual fields?  That way if one of them fails, we aren't updating `output->size` erroneously (this is probably never a problem in practice, but I still thought I would point it out).

Reply:
@clalancette even more so, we should probably be incrementing it as we copy.